services/supervisor/src/transactions.rs

```diff
@@ -178,10 +180,56 @@ impl SupervisorInterface for Supervisor {
                         .verify_config(config.params.clone())
                         .map_err(|e| (Error::MalformedConfigPropose, e))?;
                 }
+
+                ConfigChange::StartService(start_service) => {
+                    start_service.validate()?;
+
+                    if context
+                        .dispatcher_info()
+                        .get_artifact(start_service.artifact.name.as_str())
+                        .is_none()
+                    {
+                        log::trace!(
+                            "Discarded start of service {} from the unknown artifact {}.",
+                            &start_service.name,
+                            &start_service.artifact.name,
+                        );
+
+                        return Err(Error::UnknownArtifact.into());
+                    }
+
+                    if context
+                        .dispatcher_info()
+                        .get_instance(start_service.name.as_str())
+                        .is_some()
+                    {
+                        log::trace!(
+                            "Discarded start of the already running instance {}.",
+                            &start_service.name
+                        );
+                        return Err(Error::InstanceExists.into());
+                    }
+
+                    if services_to_start.contains(&start_service.name) {
+                        log::trace!(
+                            "Discarded multiple instances with the same name in one request."
+                        );
+                        return Err(Error::MalformedConfigPropose.into());
+                    }
+
+                    services_to_start.insert(&start_service.name);
+                }
             }
         }
 
         let schema = Schema::new(context.instance().name, context.fork());
+
+        // After all the checks verify that configuration number is expected one.
+        if propose.configuration_number != schema.get_configuration_number() {
```

Comment:
I think it's better to check the number *after* all the other checks, so we won't increase the number on the obviously incorrect configs.